### ISSUE TITLE
DRA drivers: helper code for checking device allocations

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#128981]
### What would you like to be added?

A DRA driver has to check in NodePrepareResources whether the devices are a) already prepared for other claims and b) really currently available.

If so, it has to refuse to prepare them again. This is a sanity check that acts as last line of defense against unintended concurrent usage of a device (case a) or race conditions. DRA tries hard to avoid such situations, but at least the ["device became unhealthy"](https://github.com/kubernetes/kubernetes/issues/128979) is inherently racy. Allocating the same device for different claims could be the result of a bug or allocation by different schedulers without [coordination](https://github.com/kubernetes/kubernetes/issues/128980#issuecomment-2500948711).

DRA driver authors might forget to implement this check. Perhaps we can move it into [k8s.io/dynamic-resource-allocation/kubeletplugin](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin) with such an API change that the driver authors are forced to do something?

### Why is this needed?

Higher quality of DRA drivers.

**Comment by olderTaoist** [comment]
can let me try?

**Comment by pohly** [comment]
/assign @olderTaoist 

The crucial part will be writing good documentation and a sane Go API.

**Comment by googs1025** [comment]
Are you still active? If not, I will take it down. @olderTaoist

**Comment by olderTaoist** [comment]
> Are you still active? If not, I will take it down. @olderTaoist 

sorry, delayed due to work, i work on now

**Comment by olderTaoist** [comment]
> If so, it has to refuse to prepare them again. This is a sanity check that acts as last line of defense against unintended concurrent usage of a device (case a) or race conditions. 
> 

I'm trying to understand what you mean, see if it's correct. `ResourceSlice` just update when kubeletplugin start, unlike the device plugin that regularly report the number of custom resources through the [ListAndWatch](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-implementation) interface, so you proposed ["device became unhealthy"](https://github.com/kubernetes/kubernetes/issues/128979) issue to resolve 
> A DRA driver has to check in NodePrepareResources whether the devices are b) really currently available.

the following problem causes same device to be assigned to different claims
> allocation by different schedulers without [coordination](https://github.com/kubernetes/kubernetes/issues/128980#issuecomment-2500948711).
>

So you want to a sanity check in `NodePrepareResources` for a) and b).

but each kubeletplugin has a different implementation for check, for example [nvidia kubeletplugin](https://github.com/NVIDIA/k8s-dra-driver) put the correspondence between claims and devices in the checkpoint file as fellow:
```
func (s *DeviceState) Prepare(ctx context.Context, claim *resourceapi.ResourceClaim) ([]*drapbv1.Device, error) {
	s.Lock()
	defer s.Unlock()

	claimUID := string(claim.UID)

	checkpoint := newCheckpoint()
	if err := s.checkpointManager.GetCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
		return nil, fmt.Errorf("unable to sync from checkpoint: %v", err)
	}
	preparedClaims := checkpoint.V1.PreparedClaims

	if preparedClaims[claimUID] != nil {
		return preparedClaims[claimUID].GetDevices(), nil
	}

	preparedDevices, err := s.prepareDevices(ctx, claim)
	if err != nil {
		return nil, fmt.Errorf("prepare devices failed: %w", err)
	}

	if err := s.cdi.CreateClaimSpecFile(claimUID, preparedDevices); err != nil {
		return nil, fmt.Errorf("unable to create CDI spec file for claim: %w", err)
	}

	preparedClaims[claimUID] = preparedDevices
	if err := s.checkpointManager.CreateCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
		return nil, fmt.Errorf("unable to sync to checkpoint: %v", err)
	}

	return preparedClaims[claimUID].GetDevices(), nil
}
```
It is difficult to implement check in`NodePrepareResources` without adding interfaces of kubeletplugin.
@pohly @googs1025 Do you have any good suggestions?

**Comment by pohly** [comment]
Yes, you have to extend the kubeletplugin interface. This forces DRA driver authors to implement additional methods, which is intentional: it makes it clear that they have to do something, which is better than the current "your NodePrepareResources implementation should do this, but no-one will notice if it was forgotten".

**Comment by pohly** [comment]
/triage accepted
/assign

https://github.com/kubernetes/kubernetes/pull/129442 can be used as a starting point. Opens:
- Config option to let DRA drivers opt out of the forced serialization.
- Contextual logging with per-gRPC-call sequence ID.

**Comment by pohly** [comment]
https://github.com/kubernetes/kubernetes/pull/130700 revises the Go API and provides everything except for forcing DRA drivers to implement claim verification. It gets called out explicitly now. I'm undecided whether that is enough.

**Comment by vishalanarase** [comment]
Looks like it's long pending issue, This needs to be triaged again as bot suggested. If it's still valid I can pick this up.

**Comment by pohly** [comment]
@vishalanarase: if you have some idea how the Go API can support this better, then we can try that out. Otherwise we should probably close.

A real-world test would be to look at the DRA driver for NVIDIA GPUs and how it currently handles this check. Would that code be better with some helper code?

**Comment by zariahazari75-dev** [comment]
I can take a look at the DRA driver for NVIDIA GPUs to see how it currently
handles device allocation checks and determine if helper code would improve
the implementation. I'll evaluate if there's a better way for the Go API to
support this and provide an update on whether we should proceed or close
the issue.

**Comment by vishalanarase** [comment]
@pohly Looking at [kubernetes-sigs/dra-driver-nvidia-gpu](https://github.com/kubernetes-sigs/dra-driver-nvidia-gpu) as the real-world test:

The GPU kubelet plugin already does both sanity checks you called out, inside `DeviceState.Prepare()`:

1. **Same claim (idempotent):** if the checkpoint has the claim in `PrepareCompleted`, prepare is a noop and returns the stored devices.
2. **Different claim (conflict):** `validateNoOverlappingPreparedDevices` walks prepared claims and fails if a requested non-admin device is already prepared for another claim UID.
3. **Actually available:** handled in driver-specific `prepareDevices` (NVML / MIG / vfio). That part is not a good fit for a generic helper.

They use the #130700 `DRAPlugin` API with `Serialize(false)` and their own prepare/unprepare + checkpoint locks.

**Would helper code make this better?** Partly. The cross-claim conflict check (~40 lines over a claim→device index with admin-access exceptions) is generic and shows up the same way in any serious driver. Hardware availability stays driver-local. The current `PrepareResourceClaims` docs call the check out, but nothing forces it. The e2e test driver only does same-claim idempotency.

A useful API shape on top of today’s helper would be:

- A small `PreparedDeviceIndex` (or similar) in `kubeletplugin` for “device already prepared for another claim?”
- Optionally a mandatory `ValidatePrepareResourceClaims` on `DRAPlugin` so drivers cannot skip the hook (availability + any driver-specific rules still live in the driver)

That would shrink NVIDIA’s conflict check without pretending to own NVML state. The earlier #129442 approach was too narrow (mostly idempotency) and used context values, this builds on the #130700 interface instead.

If that direction sounds right, I can prototype the index + interface change. If the remaining value is too small next to what production drivers already do, I’m fine closing.

**Comment by pohly** [comment]
@shivamerla, @byako: any thoughts on the above?

**Comment by shivamerla** [comment]
@pohly A helper to check whether a claim is already prepared for another non-admin claim would definitely be useful. That said, I don't think full validation should be baked in as a blanket rule, there are legitimate cases where overlapping claims need to be allowed, so driver authors should still be able to layer their own checks on top of the helper.
For example, with GPUs we need to allow overlapping claims when the consumable capacity feature is enabled (i.e., with allowMultipleAllocations set). See the discussion [here](https://github.com/kubernetes-sigs/dra-driver-nvidia-gpu/pull/1231/changes#r3495476763).
There's also a potential use case for allowing multiple claims with different opaque configs to apply to the same device. These configs wouldn't touch the hardware device itself, but would let the driver enforce its own software-level constraints per allocation.

**Comment by byako** [comment]
+1 for optional, helper-available validation, but I'm not sure it's worth it, given the lack of common / unified prepared claims cache. What I mean is - all DRA drivers have their own format of snapshot / saved state, and without leveraging it - what would the helper-available validation use?
- API calls to api-server every time a PrepareResourceClaims is called? seems unreasonably heavy / expensive
- local cache of prepared claims? Will it not be duplicate to existing saved state, that particular DRA driver already has?

**Comment by vishalanarase** [comment]
@pohly What's your opinion or final thoughts on this? So I can start on implementation.

**Comment by pohly** [comment]
Looking at the feedback above it doesn't look feasible, so let's not do it?

**Comment by vishalanarase** [comment]
Okay, Thank you.

/close
